### PR TITLE
Creating google-cloud-compat-checker

### PR DESCRIPTION
**Pull request by garrettjonesgoogle**
So far I am just bootstrapping this to check os & architecture. There is a lot more we could check, which we can add later. 

Currently, users have to build this themselves, because there have been no releases of it yet. Later, we can switch the instructions to use the maven artifact somehow. 

I intend to use this to diagnose https://github.com/GoogleCloudPlatform/google-cloud-java/issues/2266 .

### CODE REVIEW

File: google-cloud-compat-checker/src/main/java/com/google/cloud/compatchecker/GoogleCloudCompatChecker.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+      ImmutableSet.of("linux-x86_64", "osx-x86_64", "windows-x86_64");
+
+  private static class OsDetector extends Detector {
+    public void detect(Properties props, List<String> classifierWithLikes) {
+      super.detect(props, classifierWithLikes);
+    }
+
+    @Override
+    protected void log(String s) {
+      // meh
```
Comment (vam-google) [inline google-cloud-compat-checker/src/main/java/com/google/cloud/compatchecker/GoogleCloudCompatChecker.java]: Really?

File: google-cloud-compat-checker/pom.xml [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    <dependency>
+      <groupId>io.netty</groupId>
+      <artifactId>netty-tcnative-boringssl-static</artifactId>
+      <version>2.0.3.Final</version>
+    </dependency>
+    <dependency>
+      <!-- It may look weird to have a compile dependency on a plugin,
+	   but we want to execute code from it at runtime, in order
+	   to not duplicate the logic ourselves -->
+      <groupId>kr.motd.maven</groupId>
```
Comment (vam-google) [inline google-cloud-compat-checker/pom.xml]: Do we really need this strange dependency? Usually it is possible to get enough of information about the running environment by means of core java with combination of environment variables and system properties, plus, maybe some more advanced (but still built in stuff). For example:

```java
PrintStream out = new PrintStream(System.out);

out.append("\n\nSYSTEM PROPERTIES:\n\n");
System.getProperties().list(out);

//out.append("\nENVIRONMENT VARIABLES:\n\n");
//Map<String, String> envVars = System.getenv();
//for (Map.Entry<String, String> entry : envVars.entrySet()) {
//  out.append(entry.getKey()).append("=").append(entry.getValue()).append('\n');
//}

```

Among other garbage, that outputs stuff like:
```
os.name=Linux
os.version=4.4.0-83-generic
os.arch=amd64

```

File: google-cloud-compat-checker/README.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
+Google Cloud Java Compatibility Checker
```
Comment (vam-google) [inline google-cloud-compat-checker/README.md]: What do you think about renaming this into something more generic (as I believe more utility stuff can/will be added here in the future)?

File: google-cloud-compat-checker/src/main/java/com/google/cloud/compatchecker/GoogleCloudCompatChecker.java (hunk 1 above)
Comment (garrettjonesgoogle) [inline google-cloud-compat-checker/src/main/java/com/google/cloud/compatchecker/GoogleCloudCompatChecker.java]: Yeah, the implementation inside the plugin just has log.info lines in its implementation.

File: google-cloud-compat-checker/pom.xml (hunk 2 above)
Comment (garrettjonesgoogle) [inline google-cloud-compat-checker/pom.xml]: Sure, it's possible to get the information that way. *However*, the selection of netty-tcnative uses this logic. It seems both risky and wasteful to duplicate the logic, when all we want to do use get the exact result of the logic. Remember, we are trying to diagnose library loading errors here, so the more we diverge, the more likely we are to mismatch the actual library loading.

File: google-cloud-compat-checker/README.md (hunk 3 above)
Comment (garrettjonesgoogle) [inline google-cloud-compat-checker/README.md]: I would probably want other utility stuff to be in a separate module - this module is likely to have weird dependencies (for example, a compile dependency on a maven plugin package).

Comment (vam-google) [inline google-cloud-compat-checker/README.md]: Utility classes often have weird dependencies, so I believe it is ok. I just feel uncomfortable with creating ad-hoc modules which do only one thing, as it contaminates the google-cloud java library (user most probably expects a module per client, various utility modules give wrong impression that that is some real google client to some mysterious google cloud service).

**Comment by garrettjonesgoogle** [comment]
Then how about having a google-cloud-util module, and then put google-cloud-compat-checker under that?

File: google-cloud-compat-checker/README.md (hunk 3 above)
Comment (garrettjonesgoogle) [inline google-cloud-compat-checker/README.md]: Then how about having a google-cloud-util module, and then put google-cloud-compat-checker under that?

Comment (vam-google) [inline google-cloud-compat-checker/README.md]: Sounds good to me

**Comment by garrettjonesgoogle** [comment]
Moved the compat checker under google-cloud-util, PTAL

**Comment by vam-google** [comment]
I meant some (optional) description in readme of what this module is about, since for utility modules it is not that clear. It doesn't have to be called "purpose".

**Pull request merged**